src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                {
+                    // Dismiss the session.
+                    return true;
+                }
+
+                // DismissIfLastCharacterDeleted should be applied only when started with Insertion, and then Deleted all characters typed.
+                // This conforms with the original VS 2010 behavior.
+                if (_initialRoslynTriggerKind == CompletionTriggerKind.Insertion &&
+                    _data.Trigger.Reason == CompletionTriggerReason.Backspace &&
+                    _completionRules.DismissIfLastCharacterDeleted &&
+                    _filterText.Length == 0)
+                {
+                    // Dismiss the session
+                    return true;
+                }
+
+                return false;
+
+                static bool IsAfterDot(ITextSnapshot snapshot, ITrackingSpan applicableToSpan)
+                {
+                    var position = applicableToSpan.GetStartPoint(snapshot).Position;
+                    return position > 0 && snapshot[position - 1] == '.';
+                }
+            }
+
+            private IDisposable ComputeItemsToBeIncluded(out IReadOnlyList<MatchResult<VSCompletionItem>> itemsToBeIncluded)
+            {
+                var disposer = PooledListForMatchResult.CreateInstance(out var list);
+
+                // FilterStateHelper is used to decide wheter a given item should be included in the list based on the state of filter/expander buttons.
+                var filterHelper = new FilterStateHelper(_data.SelectedFilters);
+                filterHelper.LogTargetTypeFilterTelemetry(_session);
+
+                // Use a monotonically increasing integer to keep track the original alphabetical order of each item.
+                var currentIndex = 0;
+
+                // Filter items based on the selected filters and matching.
+                foreach (var item in _data.InitialSortedList)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    if (filterHelper.ShouldBeFilteredOut(item))
+                        continue;
+
+                    var roslynItem = GetOrAddRoslynCompletionItem(item);
+                    if (CompletionHelper.TryCreateMatchResult(_completionHelper, roslynItem, item, _filterText,
+                        _initialRoslynTriggerKind, FilterReason, _recentItemsManager.RecentItems, _highlightMatchingPortions, currentIndex,
+                        out var matchResult))
+                    {
+                        list.Add(matchResult);
+                        currentIndex++;
+                    }
+                }
+
+                // Sort the items by pattern matching results.
+                // Note that we want to preserve the original alphabetical order for items with same pattern match score,
+                // but `List<T>.Sort` isn't stable. Therefore we have to add a monotonically increasing integer
+                // to `MatchResult` to achieve this.
```

Comment:
move this comment up to the comment on currentIndex.  Current index shoudl exlpain that it is needed so we can ensure teh sort below is stable.